### DISCUSSION TITLE
Release 0.7.0 (1.21.30.23): Personalization Update

### DISCUSSION THREAD

**Discussion by m-brisch** [Mojang/minecraft-editor#192] [Announcements]
_[Read the previous release notes here](https://github.com/Mojang/minecraft-editor/discussions/188)_

---

## Summary

Hello! Happy Preview day 😊 This week we are officially closing the v0.6.X semantic iterations and considering our work towards v0.7 done. Next update, we will begin 0.7.X cycles towards our next v0.8 release, which will be focused on improving the core building experience by expanding the number of built in tools, improving performance + stability, and building on features like the hotbar and structure panel that were introduce in v0.7!

Read on for more details on this week’s exciting updates, as well as an all-up summary of everything that was done during the v0.6.X iterations.

As a reminder of Editor API support:

- Editor Extensions [overview article](https://learn.microsoft.com/en-us/minecraft/creator/documents/editorextensionsintroduction) and [API reference documentation](https://aka.ms/EditorAPI) on learn.microsoft.com
- [Starter Kit](https://github.com/Mojang/minecraft-editor-extension-starter-kit): Contains the compiler tools and various libraries required to get started making Editor Extensions
- [Samples](https://github.com/Mojang/minecraft-editor-extension-samples) created by our team to jump-start your own Extension creations

Additionally, if you're new to the Editor GitHub this week, we encourage you to join the Bedrock Add-Ons Discord [#Editor channel](https://discord.gg/TZnQKFwGpt)!

---

### Structures Panel

The new Structures Panel is the hub for keeping structures that you will be able to move back-and-forth from the clipboard. This means you can quickly build complex scenes by stamping prebuilt structures instead of relying on copy/paste or rebuilding them one-by-one. The Structures Panel is synchronized so if you'll be able to see and use structures that are saved by other players in your world. Right now the pane is limited and read only, but in future updates we will be both expanding the capabilities within the panel and adding dedicated tools that can pull from these structures. You can add a shortcut to the Structures Panel via the action bar. We're early on our journey here so we'd love to hear what features you'd want to see in Bedrock Add-Ons Discord [#Editor channel](https://discord.gg/TZnQKFwGpt)!

Easily create a structure by:

1. Select an area in the world
2. Hit `CTRL+C` to copy it to your clipboard
3. Open the Structures Panel from the action bar
4. Name your structure and hit `New from Clipboard`

https://github.com/user-attachments/assets/928c0d33-6c90-401f-93f0-8e167802cbe3

Then when you want to use your structures:

1. Open the Structures Panel from the action bar
2. Select the structure you want to use
3. Hit 'Copy to Clipboard'
4. Hit `CTRL+SHIFT+V` to begin Paste Preview

https://github.com/user-attachments/assets/0bee64b4-a9d1-4f2d-bf75-4c5b285d1563

---

### Navigation Panel

We're introducing a new Navigation Panel to make it easier to navigate your worlds. At the top is a toggle to show a birds-eye view mini-map which will teleport you to that coordinate on click. Your player's location and orientation will be displayed at the center of the map. At the bottom you can add markers to locations in your world and they will show up as colored flags within your mini-map. You can find the Navigation Panel under `World Options->Navigation Panel` or add a shortcut in the action bar.

![image](https://github.com/user-attachments/assets/3f2c8cc1-f8a6-426a-91d5-e7a2d892cd46)

---

### Hardcore Mode

We added support for hardcore mode in the export pane (`File->Export`). Note that some settings like cheats and difficult are overridden and restricted when hardcore mode is enabled.

---

### Full Change Log

- Added Structure Panel action bar item for managing structures stored in the project.
  - Added Editor Structure Manager API to interface with it.
- Added Navigation Panel which adds a mini-map to teleport around your world and add labels for key points
- Added Hardcore mode as an option in the export pane
- Added `getSelectedToolId` and `setSelectedToolId` functions to `IModalToolContainer` API
- Fixed a bug where Editor operations could exhaust all available ticking areas causing all future operations to fail.
- Fixed a bug where sometimes `IModalTool.onModalToolActivation` API event was called in the wrong order
- Fixed a bug which could sometimes cause brush masks to fail to load

---

## v0.6.X Updates Aggregation

Here is the list of updates that have occurred in the Editor since v0.6.1 on April 17, 2024 to v0.7.0 on August 15, 2024.

### Major Features

<details><summary>Paste Preview</summary>

Added a new Paste Preview tool which supports preview visualization of the clipboard contents and enables rotate and mirror.
![pastepreviewvis](https://github.com/Mojang/minecraft-editor/assets/29804144/45dc5ddc-d005-4638-b45b-3b5c2a1bfa04)

Additional functionality has also been added to keyboard bindings. We want to add these paradigms to other tools, so we'd love your feedback on what you find useful and brought to other tools!

- SHIFT+CTRL+V - Activate Paste Preview Mode (if there's content in the clipboard)
- CTRL+C - Copy Selection to Clipboard
- CTRL+V - Immediate paste to cursor position (no preview)
- Arrows/Pg UP/Pg DN - manually nudge the 3D block cursor in a camera relative direction
- CTRL + Arrows/Pg UP/Pg DN - Manually nudge the pivot location inside the bounds of the clipboard item, in a camera relative direction
- SHIFT+CTRL + Arrows/Pg UP/Pg DN - Manually nudge the placement offset, in a camera relative direction
- CTRL+D or ESC - Deselect Preview Mode
- Left Mouse Click, ENTER or CTRL+F - Place paste preview in the world at the 3D block cursor position
- CTRL + Mouse Click - Remove blocks in the shape of the clipboard item (Note: Current implementation will only use the bounding box... complex shape removal is still in development)
- B - Cursor Block Mode
- F - Cursor Adjacent (Face) Mode
- R - Rotate Cycle CW (0, 90, 180, 270) around Y axis at pivot location
- SHIFT+R - Rotate Cycle CCW (270, 180, 90, 0) around Y at pivot location
- X - Mirror X Axis around pivot
- Z - Mirror Z Axis around pivot
- C - Toggle Cursor Control Mode between "Keyboard & Mouse" or "Keyboard Only"
- D - Change Cursor Control mode to "Fixed Distance"
- Scroll Wheel - Adjust Fixed Cursor distance (only when Fixed distance is enabled)

</details>

<details><summary>Deferred Lighting Settings</summary>

We added a tool to modify the deferred lighting settings directly within Editor so you can easily fine-tune your world's experience. You can open the tool panel by going to `View => Deferred Lighting Settings` in the top menu. When you're done configuring your world you can hit the `Export to clipboard` button at the bottom of the page which you can then paste directly into your resource pack.

https://github.com/Mojang/minecraft-editor/assets/29804144/3a79dee2-d0c2-498c-b226-3f262436d9ab

https://github.com/user-attachments/assets/62718c3a-3d98-47c5-b998-654d0d1d0242

For more information about Deferred Lighting, check out the [learn page](https://learn.microsoft.com/en-us/minecraft/creator/documents/deferredlighting?view=minecraft-bedrock-stable).

</details>

<details><summary>Custom Input Mapping</summary>

We added a tool to allow you to remap the shortcuts used by Editor and all Editor Extensions! From the Action Bar hit `+` and then `Input Mapping` to add the shortcut to your Action Bar. Then hit `Input Mapping` to open the panel.

- Key bindings can be reconfigured using the shortcut field, which detects key presses to set the binding. Pending changes will highlight the text until saved or cleared. Closing the modal reverts any pending changes.
- Right-clicking on a selected shortcut field clears the binding, or clear button next to the field can be used.
- Added a toggle for exclusive (InputModifier.None in API) key actions to support advanced input scenarios.
- Modified bindings can be restored by clicking restore buttons for the binding, active context, and all modified bindings.
- Saved bindings are accessible across different editor projects.

![image](https://github.com/Mojang/minecraft-editor/assets/29804144/ea0a348e-ba50-4ffd-97da-921d997aaf75)

</details>

<details><summary>Entity Transactions</summary>

We updated the summon tool to use `TransactionManager` to track the addition and deletion of entities. This means you're now able to use undo/redo to more easily create entity rich environments.

</details>

<details><summary>Client Widget System</summary>

We added a new set of client widgets to help make your Editor Extensions more expressive! We're super excited to see what you'll build with them. If you have ideas on more widgets you need for your extension, let us know either here or on discord!

https://github.com/user-attachments/assets/841b3eea-9f40-4cf3-98cd-f813dc827c9a

</details>

<details><summary>Export Pane</summary>

We added a new panel to our export flow so you're able to customize the world settings before exporting to file
![image](https://github.com/user-attachments/assets/7f906298-d6e6-4d0b-a336-330f3066d7c2)

</details>

<details><summary>Brush Tool + masking support</summary>

We made several improvements to the brush pipeline which should make things much more performant. There's a new `Single Block` brush shape for detailed work. We also added `Masking Mode` which allows you to enumerate types of blocks you want the brush to effect or ignore.

Here's an example of using the replace mask to only apply the brush to the `acacia_leaves`:

https://github.com/user-attachments/assets/86ba9ae9-9964-4ce8-a8c4-749ed99ed72e

</details>

<details><summary>Ruler Tool</summary>

We added a new ruler tool to help plan out builds and show what the new client widget system can do.

https://github.com/user-attachments/assets/4be57a82-b710-4b1c-bd82-6e50bfe1b4b2

</details>

<details><summary>Structure Manager</summary>

The new Structures Panel is the hub for keeping structures that you will be able to move back-and-forth from the clipboard. This means you can quickly build complex scenes by stamping prebuilt structures instead of relying on copy/paste or rebuilding them one-by-one. The Structures Panel is synchronized so if you'll be able to see and use structures that are saved by other players in your world. Right now the pane is limited and read only, but in future updates we will be both expanding the capabilities within the panel and adding dedicated tools that can pull from these structures. You can add a shortcut to the Structures Panel via the action bar. We're early on our journey here so we'd love to hear what features you'd want to see in Bedrock Add-Ons Discord [#Editor channel](https://discord.gg/TZnQKFwGpt)!

Easily create a structure by:

1. Select an area in the world
2. Hit `CTRL+C` to copy it to your clipboard
3. Open the Structures Panel from the action bar
4. Name your structure and hit `New from Clipboard`

https://github.com/user-attachments/assets/928c0d33-6c90-401f-93f0-8e167802cbe3

Then when you want to use your structures:

1. Open the Structures Panel from the action bar
2. Select the structure you want to use
3. Hit 'Copy to Clipboard'
4. Hit `CTRL+SHIFT+V` to begin Paste Preview

https://github.com/user-attachments/assets/0bee64b4-a9d1-4f2d-bf75-4c5b285d1563

</details>

<details><summary>Navigation Panel</summary>

We're introducing a new Navigation Panel to make it easier to navigate your worlds. At the top is a toggle to show a birds-eye view mini-map which will teleport you to that coordinate on click. Your player's location and orientation will be displayed at the center of the map. At the bottom you can add markers to locations in your world and they will show up as colored flags within your mini-map. You can find the Navigation Panel under `World Options->Navigation Panel` or add a shortcut in the action bar.

![image](https://github.com/user-attachments/assets/3f2c8cc1-f8a6-426a-91d5-e7a2d892cd46)

</details>

<details><summary>Simple Tool Wrapper</summary>

We added a `SimpleToolWrapper` framework to simplify and reduce the amount of code required to make an Editor tool. Check out `simple-empty` and `simple-locate-biome` in our [Editor Extension Starter Kit](https://github.com/Mojang/minecraft-editor-extension-starter-kit/) later today to see examples on how to get started (or just to use the locate tool, it's pretty handy).

</details>

<details><summary>Updated Property Pane API</summary>
Updated the Property Pane API to use a new type safe `IObservable` pattern. Our goal with these changes is to improve the usability and type safety when building panels inside of extensions. In addition to improved performance we also added some addition functionality to each API to improve what you're able to express within an extension. We have updated our [Editor Samples](https://github.com/Mojang/minecraft-editor-extension-samples) to help get you started with the new API changes. We've also maintained the previous API schema with `add..._deprecated` functions, but encourage everyone to upgrade to the new pattern.

</details>

<details><summary>Attach Debugger</summary>

To make it easier to debug your own Editor Extensions we added a button you can add to your Action Bar which will attempt to connect to the Minecraft script debugger on port 19144.

For more information about the Minecraft script debugger, check out the [GitHub page](https://github.com/Mojang/minecraft-debugger)

</details>

### Full Changelog

<details>
<summary>Click to expand</summary>

- Implemented a Simple Tool Wrapper framework to simplify the boilerplate implementation of setting up an editor tool, and manage the window visibility and lifecycle of the tool components.
- Editor Export API can now export a playable world
- Fixed the UI layout drawer content to respond properly to changes in the visibility of the PropertyPane API
- Fixed bug for when a player switched to Test World while in the Nether/End the worlds dimension incorrectly changed to the Overworld
- Updated [Editor Extension Starter Kit](https://github.com/Mojang/minecraft-editor-extension-starter-kit/) to include samples for custom brushes and Simple Tool Wrapper
- Fixed bug that caused crashes during deletion of Script Widget Groups
- Selection volumes can no longer be created when they would exceed the world height limits
- When filling up a volume and dragging the volume during the filling operation the filling volume would also change. Now the filling operation happens in the original volume
- Fixed a bug where camera collision caused third-person view to display inside the player's head in Crosshair mode
- Added addTable API
- Fixed GIF asset support for addImage API
- Fixed a bug where the viewport did not scale properly with changes in UI or font scaling
- Resolved tooltip positioning and content size issues
- Fixed a bug where clicking repeatedly on the "Uniform" checkbox in brush tool settings could cause a crash
- Add _enabled_ property to _IMenu_ and _IMenuCreationParams_
- Added customization and pagination support for the Action bar. Action selector can be opened from plus button to add new items or from bottom right arrow of existing items to modify them
- Updated documentation for SimpleTool classes
- Snow layer now falls when placed with a brush
- Remove Hardcore mode option when creating an editor project
- Updated Selection, Brush, Line, Paste, Delete, Fill tool to load bounds before applying changes. This should address issues where changes were not being correctly applied to the world
- Added IActionBar API to manage scripted IActionBarItem
- Added property actionBar: IActionBar to IPlayerUISession API which represents quick action items in the action bar selector UI
- Added addColorPicker method to IPropertyPane API to display a color picker property pane item for modifying RGBA values
- Removed Keybinding for ‘D’ from Paste Preview which blocked movement right. Fixed Distance mode can be selected using the side-panel
- Fixed missing CTRL+X key binding in Editor
- Expanded the Block Palette Service API to allow management of Block Palettes and Block Palette Items
- Added radial drag to number input fields, allowing value changes with mouse drag movement
- Introduced a new script brush manager method getBrushVolume which, given a set of brush pipeline operations, builds an absolute compound block volume for the brush
- Fixed bug where small movements in the brush tool caused incorrect cursor tracking
- Added settings panel to modify atmospheric, global illumination, and color grading settings for deferred lighting (requires deferred lighting resource pack and settings to be enabled)
  - Known issue: shadows are cast from the UI elements. You can toggle crosshair mode via CTRL+TAB as a workaround
  - Known issue: paste preview (CTRL+SHIFT+V) does not render properly when deferred lighting is enabled.
- Added Input Mappings action bar item that opens a modal displaying all available key bindings and input contexts.
  - Key bindings can be reconfigured using the shortcut field, which detects key presses to set the binding. Pending changes will highlight the text until saved or cleared. Closing the modal reverts any pending changes.
  - Right-clicking on a selected shortcut field clears the binding, or clear button next to the field can be used.
  - Added a toggle for exclusive (InputModifier.None in API) key actions to support advanced input scenarios.
  - Modified bindings can be restored by clicking restore buttons for the binding, active context, and all modified bindings.
  - Saved bindings are accessible across different editor projects.
- Extended game options to include more export options as well as applying options to export level data.
- Added an optional KeyBindingInfo argument to the key binding registration API functions to assign additional information to shortcuts.
- Added new BlockList UI element to Property Pane API.
- Added inputContextId and inputContextLabel optional properties to ModalToolCreationParameters API to represent user defined modal tool identifiers for modal tools.
- Added interface KeyBinding and modified key binding registration API methods to take this object.
- Dropdown menus will attempt to open in the opposite direction when they extend beyond the window bounds.
- Added variant: ColorPickerVariant optional property to IPropertyItemOptionsColorPicker API to support different types of color picker property items. Changed the default variant to respect the pane layout. Moved gradient selector for Default and Inline variants to a popup, accessible by clicking the selected color box.
- Updated set of entities that appear within the Editor picker
- Updated set of blocks that appear within the Editor picker
- Introduced a new filter "is_navigating" that checks to see if the entity is currently pathfinding. This requires a "minecraft:navigation" component
- Updated set of blocks + entities that appear within the Editor picker
- Added 'Attach Debugger' ActionBarItem shortcut for attaching to the scripting debugger
- Implemented single entity transaction payload for entity delete/create
- Implemented Undo/Redo of single entity creation and deletion
- Added isFloat to support floating point values in addNumber Editor Property Pane API
- Added copy to clipboard button to log panel messages
- Added localized field names into the Deferred Lighting property editor dropdown
- Added Global Illumination settings to the Deferred Lighting property editor
- Fixed bug that caused visual artifacts when enabling Deferred Lighting in Editor
- Removed fallback display text properties from API in favor of single localized property.
  - Added labelproperty to IDropdownItem API, removed displayAltText and displayStringId properties
  - Added labelproperty to IMenuCreationParams API, removed name and displayStringId properties
  - Added titleproperty to IPropertyItemOptions and IPropertyItem APIs, removed titleAltText and titleStringId properties
  - Added title and tooltip properties to ModalToolCreationParameters API, removed titleAltText, titleStringId, tooltipAltText and tooltipStringId properties
- Updated block hotbar to have non-air defaults
- Added persistence for the default hotbar
- Added "ExtensionContext.settings.theme" which will represent the In-World UI Theme
  - This feature currently has three functions: resolveColorKey(string): Color, setCurrentTheme(string): void, and getThemeList(): string[]
  - There is a single "minecraft:default" theme
  - Future work will add the ability to create, modify, and delete custom themes but the default themes will not be modifiable
- Updated subset of PropertyPane API to use IObservable pattern.
  - Added an alternative Property Pane addVector3API that leverages type safe IObservable Observables can be created using makeObservable API
    - Added interfaces IVector3PropertyItemand IVector3PropertyItemOptions for the new property item
    - Renamed function to create property bag Vector3 to addVector3_deprecated
  - Added an alternative Property Pane addBoolAPI that leverages type safe IObservableProp value API
    - Added interfaces IBoolPropertyItemand IBoolPropertyItemOptions for the new property item
    - Renamed function to create property bag bool item to addBool_deprecated
  - Converted Property Pane addText API to use IObservableProp
  - Added interfaces IButtonPropertyItemand IButtonPropertyItemOptions for the button property item
    - IPropertyPane addButtonAPI can now take a plain function
    - Removed ButtonVarianttype, and added ButtonPropertyItemVariant enum API
  - Renamed EDITOR_PANE_PROPERTY_ITEM_TYPEenum API to PropertyItemType
  - Converted property pane addImageAPI to support IObservable. It will also be possible to use custom image formats like icon and block with ImageResourceData as value
    - Added onClickfunction and alignment property to IImagePropertyItemOptions API
- Revamped Client Widget System
  - Refactored the Client Widget system to use components for base widget functionality
  - Added Editor Client Renderer Helper Service to aid with Editor in-world widget rendering
  - Changed Gizmo movement to be completely free moving and snap to grid on release
  - Added new widget components for Entities, Text, Gizmos, Height Guide, and basic Render Primitives
- Additional updates to the PropertyPane API
  - Removed IPropertyPane addBlockPicker and addEntityPicker methods in favor of addComboBox API that leverages type safe IObservableProp<string> value
  - Added an alternative Property Pane addNumberAPI that leverages type safe IObservableProp<number> value API
  - Updated theme colors for input fields, and center aligned number input field values
  - Added an alternative Property Pane addDropdownAPI that leverages type safe IObservableProp<number> value
  - Fixed theming and content alignment issues for the dropdown UI elements
  - Added an alternative Property Pane addColorPicker API that leverages type safe IObservableProp<RGBA> value API
  - Added an alternative Property Pane addString API that leverages type safe IObservableProp<string> value API
- Fixed missing Summon Tool icon
- Added the following functions to "ExtensionContext.settings.theme": addNewTheme(name: string): void, deleteTheme(name: string): void, getCurrentTheme(): string, and updateThemeColor(name: string, key: ThemeSettingsColorKey, newColor: minecraftserver.RGBA): void. These functions will throw if "minecraft:default" or any other built-in theme is passed to them. updateThemeColor and deleteTheme will also throw if the theme does not exist
- Added block image support to block picker modal
- Fixed a bug where typing certain characters in the block or entity picker caused an error
- Added validation to prevent adding already existing block in a probability palette item
- Added `Water` dropdown inside the `Deferred Lighting Settings` panel.
- Added `Export` pane inside the file menu to allow customizing world settings before exporting to file.
- Updated Brush Tool with additional features and improved performance.
  - Added `Single Block` brush shape to allow more fine grain editing.
  - Added `Masking Mode` to customize how brushes interact with the world.
    - `Replace` will create a list of blocks that will be replaced when brushing. Blocks not in this list will not be impacted.
    - `Mask` will create a list of blocks that will be masked/ignored when brushing. Only blocks not in the list will be impacted.
- Added `IRootPropertyPane` and `ISubPanePropertyItem` interfaces to improve property pane API usability
  - Renamed `createPropertyPane` and `removePropertyPane` functions for `IPropertyPane` to `createSubPane` and `removeSubPane`.
  - Added action button support for root property pane header.
  - Added visual customization support for sub panes.
- Added Ruler tool for measuring distances with new client widget framework.
- Added constraint on the weight value in a Block palette item. Valid values are in the range [1,100].
- Added missing Widget binding export

</details>

---

Thank you for being here! You're the best! ❤

**Comment by MetalManeMc** [comment]
@m-brisch Full Change Log section seems to have been copied from 0.6.11 by mistake.

**Reply by m-brisch** [reply]
Ah good catch, should be updated with the correct info :)
